substrate/runtime/contract/src/exec.rs

```diff
@@ -0,0 +1,257 @@
+				}
+			};
+
+			(exec_result, nested.overlay.into_change_set())
+		};
+
+		self.overlay.commit(change_set);
+
+		Ok(CallReceipt {
+			return_data: exec_result.return_data,
+		})
+	}
+
+	pub fn create(
+		&mut self,
+		caller: T::AccountId,
+		endownment: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		ctor: &[u8],
+		_data: &[u8],
+	) -> Result<CreateReceipt<T>, &'static str> {
+		let create_base_fee = <Module<T>>::create_base_fee();
+		if gas_meter.charge(create_base_fee).is_out_of_gas() {
+			return Err("not enough gas to pay base create fee");
+		}
+
+		let dest = T::DetermineContractAddress::contract_address_for(ctor, &self.self_account);
+		if <CodeOf<T>>::exists(&dest) {
+			// TODO: Is it enough?
+			return Err("contract already exists");
+		}
+
+		let change_set = {
+			let mut overlay = OverlayAccountDb::new(&self.overlay);
+
+			if endownment > T::Balance::zero() {
+				transfer(
+					gas_meter,
+					true,
+					&self.self_account,
+					&dest,
+					endownment,
+					&mut overlay,
+				)?;
+			}
+
+			let mut nested = ExecutionContext {
+				overlay: overlay,
+				self_account: dest.clone(),
+				depth: self.depth + 1,
+			};
+			let exec_result = {
+				vm::execute(
+					ctor,
+					&mut CallContext {
+						ctx: &mut nested,
+						_caller: caller,
+					},
+					gas_meter,
+				).map_err(|_| "vm execute returned error while create")?
+			};
+
+			nested.overlay.set_code(&dest, exec_result.return_data().to_vec());
+			nested.overlay.into_change_set()
+		};
+
+		self.overlay.commit(change_set);
+
+		Ok(CreateReceipt {
+			address: dest,
+		})
+	}
+}
+
+fn transfer<T: Trait>(
+	gas_meter: &mut GasMeter<T>,
+	contract_create: bool,
+	transactor: &T::AccountId,
+	dest: &T::AccountId,
+	value: T::Balance,
+	overlay: &mut OverlayAccountDb<T>,
+) -> Result<(), &'static str> {
+	let would_create = overlay.get_balance(transactor).is_zero();
+
+	let fee: T::Balance = if contract_create {
+		<Module<T>>::contract_fee()
+	} else {
+		if would_create {
+			<staking::Module<T>>::creation_fee()
+		} else {
+			<staking::Module<T>>::transfer_fee()
+		}
+	};
+
+	if gas_meter.charge_by_balance(fee).is_out_of_gas() {
+		return Err("not enough gas to pay transfer fee");
+	}
+
+	let from_balance = overlay.get_balance(transactor);
+	if from_balance < value {
+		return Err("balance too low to send value");
+	}
+	if would_create && value < <staking::Module<T>>::existential_deposit() {
+		return Err("value too low to create account");
+	}
+	if <staking::Module<T>>::bondage(transactor) > <system::Module<T>>::block_number() {
+		return Err("bondage too high to send value");
+	}
+
+	let to_balance = overlay.get_balance(dest);
+	if to_balance + value <= to_balance {
+		return Err("destination balance too high to receive value");
+	}
+
+	if transactor != dest {
+		overlay.set_balance(transactor, from_balance - value);
+		overlay.set_balance(dest, to_balance + value);
+	}
+
+	Ok(())
+}
+
+struct CallContext<'a, 'b: 'a, T: Trait + 'b> {
```

Comment:
It actually depends on the usage and in fact in some previous versions this code used only single `'a`. The reason why `'b` was introduced is because lifetime inference worked badly and erroneously introduced false lifetime dependencies. I don't remember what exactly, but it was something like that I wasn't able to use `overlay` even if all borrowing refs are not longer in scope.

Reply:
If you try to use only `'a`, then you will get this error:

```
error[E0505]: cannot move out of `nested.overlay` because it is borrowed50/151: substrate-runtime-contract
  --> substrate/runtime/contract/src/exec.rs:97:18
   |
85 |                         ctx: &mut nested,
   |                                   ------ borrow of `nested` occurs here
...
97 |             (exec_result, nested.overlay.into_change_set())
   |                           ^^^^^^^^^^^^^^ move out of `nested.overlay` occurs here

error[E0499]: cannot borrow `nested.overlay` as mutable more than once at a time
   --> substrate/runtime/contract/src/exec.rs:156:4
    |
149 |                         ctx: &mut nested,
    |                                   ------ first mutable borrow occurs here
...
156 |             nested.overlay.set_code(&dest, exec_result.return_data().to_vec());
    |             ^^^^^^^^^^^^^^ second mutable borrow occurs here
157 |             nested.overlay.into_change_set()
158 |         };
    |         - first borrow ends here

error[E0505]: cannot move out of `nested.overlay` because it is borrowed
   --> substrate/runtime/contract/src/exec.rs:157:4
    |
149 |                         ctx: &mut nested,
    |                                   ------ borrow of `nested` occurs here
...
157 |             nested.overlay.into_change_set()
    |             ^^^^^^^^^^^^^^ move out of `nested.overlay` occurs here

error: aborting due to 3 previous errors
```